dotnet/src/dotnetcore/GxClasses.Web/Middleware/GXRouting.cs

```diff
@@ -0,0 +1,525 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Text.Json.Serialization;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using GeneXus;
+using GeneXus.Application;
+using GeneXus.Configuration;
+using GeneXus.Http;
+using GeneXus.Metadata;
+using GeneXus.Procedure;
+using GeneXus.Utils;
+using log4net;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+
+namespace GxClasses.Web.Middleware
+{
+	public class GXRouting : IGXRouting
+	{
+
+		static readonly ILog log = log4net.LogManager.GetLogger(typeof(IGXRouting));
+
+		public static string VirtualPath = string.Empty;
+		public static string LocalPath = Directory.GetCurrentDirectory();
+		public static string ContentRootPath;
+		static char[] urlSeparator = { '/', '\\' };
+		const char QUESTIONMARK = '?';
+		public static string UrlTemplateControllerWithParms;
+
+		//Azure Functions
+		public bool AzureRuntime;
+		public static string AzureFunctionName;
+
+		const string SDSVC_METHO_SUFFIX = "DL";
+		static Regex SDSVC_PATTERN = new Regex("([^/]+/)*(sdsvc_[^/]+/[^/]+)(\\?.*)*");
+
+		const string PRIVATE_DIR = "private";
+		public Dictionary<string, string> servicesPathUrl = new Dictionary<string, string>();
+		public Dictionary<String, Dictionary<string, SingleMap>> servicesMap = new Dictionary<String, Dictionary<string, SingleMap>>();
+		public Dictionary<String, Dictionary<Tuple<string, string>, String>> servicesMapData = new Dictionary<String, Dictionary<Tuple<string, string>, string>>();
+		public Dictionary<string, List<string>> servicesValidPath = new Dictionary<string, List<string>>();
+		public string restBaseURL;
+
+		public GXRouting(string baseURL)
+		{
+			restBaseURL = baseURL;
+			ServicesGroupSetting();
+			ServicesFunctionsMetadata();
+		}
+
+		static public List<ControllerInfo> GetRouteController(Dictionary<string, string> apiPaths,
+											Dictionary<string, List<string>> sValid,
+											Dictionary<string, Dictionary<string, SingleMap>> sMap,
+											Dictionary<string, Dictionary<Tuple<string, string>, string>> sMapData,
+											string basePath, string verb, string path)
+		{
+			List<ControllerInfo> result = new List<ControllerInfo>();
+			string parms = string.Empty;
+			string method = string.Empty;
+			GXLogging.Debug(log, $"GetRouteController basePath:{basePath} verb:{verb} path:{path}");
+			try
+			{
+				if (!string.IsNullOrEmpty(path))
```

Comment:
Suggestion: Prefer early return versus if statement.
We could exit early if path is empty. This would make the code more readable with one less indentation.